### PR TITLE
mongo: valid json logging

### PR DESCRIPTION
**Pull request by danielhochman**

### CODE REVIEW

File: source/common/common/utility.cc [hunk 1]
```diff
@@ -117,6 +117,35 @@ std::string StringUtil::subspan(const std::string& source, size_t start, size_t
   return source.substr(start, end - start);
 }
 
+std::string StringUtil::escape(const std::string& source) {
+  std::string ret;
+
+  // Prevent unnecessary allocation by allocating 2x original size.
+  ret.reserve(source.length() * 2);
+  for (size_t i = 0; i < source.length(); i++) {
```
Comment (ccaraman) [inline source/common/common/utility.cc]: for(char& c : str) ?

File: source/common/mongo/codec_impl.cc [hunk 2]
```diff
@@ -44,9 +43,10 @@ bool GetMoreMessageImpl::operator==(const GetMoreMessage& rhs) const {
 }
 
 std::string GetMoreMessageImpl::toString(bool) const {
-  return fmt::format("[GET_MORE id={} response_to={} collection='{}' return={} cursor={}]",
-                     request_id_, response_to_, full_collection_name_, number_to_return_,
-                     cursor_id_);
+  return fmt::format(
+      R"EOF({{"opcode": "OP_GET_MORE", "id": {}, "response_to": {}, "collection": "{}", "return": {}, )EOF"
```
Comment (ccaraman) [inline source/common/mongo/codec_impl.cc]: general q: why is the double { needed?

File: source/common/common/utility.cc [hunk 3]
```diff
@@ -117,6 +117,35 @@ std::string StringUtil::subspan(const std::string& source, size_t start, size_t
   return source.substr(start, end - start);
 }
 
+std::string StringUtil::escape(const std::string& source) {
```
Comment (ccaraman) [inline source/common/common/utility.cc]: Did you forget to include the tests for this method?

File: source/common/mongo/codec_impl.cc (hunk 2 above)
Comment (danielhochman) [inline source/common/mongo/codec_impl.cc]: fmt uses `{}` to find the field to replace. `{{` is effectively escaping `{` so fmt doesn't start trying to parse

Comment (ccaraman) [inline source/common/mongo/codec_impl.cc]: Thanks I didn't know that.

File: source/common/common/utility.cc [hunk 4]
```diff
@@ -117,6 +117,34 @@ std::string StringUtil::subspan(const std::string& source, size_t start, size_t
...
+  std::string ret;
+
+  // Prevent unnecessary allocation by allocating 2x original size.
+  ret.reserve(source.length() * 2);
+  for (const auto& c : source) {
```
Comment (ccaraman) [inline source/common/common/utility.cc]: nit: use char for clarity

File: test/common/mongo/codec_impl_test.cc [hunk 5]
```diff
@@ -288,4 +305,22 @@ TEST_F(MongoCodecImplTest, InvalidMessage) {
...
+  query.flags(0x4);
+  query.fullCollectionName("test");
+  query.numberToSkip(20);
+  query.numberToReturn(-1);
+  query.query(Bson::DocumentImpl::create()->addString("string_need_esc", "{\"foo\": \"bar\n\"}"));
+
+  EXPECT_EQ(
+      R"EOF({"opcode": "OP_QUERY", "id": 1, "response_to": 1, "flags": "0x4", )EOF"
+      R"EOF("collection": "test", "skip": 20, "return": -1, "query": )EOF"
+      R"EOF({"string_need_esc": "{\"foo\": \"bar\n\"}"}, "fields": {}})EOF",
```
Comment (ccaraman) [inline test/common/mongo/codec_impl_test.cc]: Should the string be escaped?

Comment (danielhochman) [inline test/common/mongo/codec_impl_test.cc]: it is. these are raw strings as denoted by the `R`.

**Pull request merged**